File: cascadingFailuresTaskMigration/src/main/java/MPFTM/AdLeadersReplace.java
package MPFTM;

import input.Group;
import input.Robot;
import org.jgrapht.alg.scoring.BetweennessCentrality;
import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.HashMap;
import java.util.List;
import java.util.Set;

public class AdLeadersReplace {
    private DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph;
    private HashMap<Integer, Group> idToGroups;
    private HashMap<Integer,Robot> idToRobots;
    public AdLeadersReplace(HashMap<Integer, Group> idToGroups, HashMap<Integer, Robot> idToRobots, DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph) {
        this.idToGroups = idToGroups;
        this.idToRobots = idToRobots;
        this.arcGraph = arcGraph;
    }
    public void run() {
        for (Integer groupId : idToGroups.keySet()) {
            Group group = idToGroups.get(groupId);
            if (group.getLeader().getFaultA()==1) {
                replace(group);
            }
        }
    }

    private void replace(Group group) {
        List<Robot> adLeaders = group.getAdLeaders();
        DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> subGraph = new DefaultUndirectedWeightedGraph<>(DefaultWeightedEdge.class);
        Set<Integer> robotIdSet = group.getRobotIdInGroup();
        for (Integer robotId : robotIdSet) {
            subGraph.addVertex(robotId);
            Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(robotId);
            for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
                Integer target = subGraph.getEdgeTarget(defaultWeightedEdge);

                if (target.equals(robotId)) {
                    continue;
                }

                if (group.getGroupId()!=idToRobots.get(target).getGroupId()) {
                    continue;
                    //不属于这一层的节点（比如由leader节点与其他lead而节点相连节点），不要加到子图中
                }
                subGraph.addVertex(target);
                //把这一层的节点加入到子图中
                if(!subGraph.containsEdge(robotId,target)){
                    //去掉重复的边,不包含这个边，则加入这个边
                    DefaultWeightedEdge temp = subGraph.addEdge(robotId,target);
                    subGraph.setEdgeWeight(temp,arcGraph.getEdgeWeight(defaultWeightedEdge));
                }
            }
        }
        //计算子图的介数中心性
        BetweennessCentrality<Integer, DefaultWeightedEdge> betweennessCentrality = new BetweennessCentrality<>(subGraph);
        //这里的I是衡量介数中心性的I
        Robot replaceLeader = adLeaders.get(0);
        Double maxIscore = -1.0;
        for (Robot adLeader : adLeaders) {
            Double Iscore =  (betweennessCentrality.getVertexScore(adLeader.getRobotId())+1)/(1-(1-adLeader.getFaultA())*(1-adLeader.getFaultO()));
            if (Iscore>maxIscore) {
                replaceLeader = adLeader;
                maxIscore = Iscore;
            }
        }
        group.setLeader(replaceLeader);
        adLeaders.remove(replaceLeader);
        group.setAdLeaders(adLeaders);
    }
}

File: FAO_case_study/algorithms/MPFTM/ad_leaders_replace.py
import networkx as nx
from typing import Dict, List
from input.group import Group
from input.robot import Robot


class AdLeadersReplace:
    def __init__(self, id_to_groups: Dict[int, Group], id_to_robots: Dict[int, Robot],
                 arc_graph: nx.Graph):
        self.id_to_groups = id_to_groups
        self.id_to_robots = id_to_robots
        self.arc_graph = arc_graph

    def run(self):
        """Replace failed leaders with backup leaders."""
        for group_id in self.id_to_groups.keys():
            group = self.id_to_groups[group_id]
            if group.leader.fault_a == 1:
                self._replace(group)

    def _replace(self, group: Group):
        """Replace leader with best backup leader."""
        ad_leaders = group.ad_leaders

        # Create subgraph
        sub_graph = nx.Graph()
        robot_id_set = group.robot_id_in_group

        for robot_id in robot_id_set:
            sub_graph.add_node(robot_id)
            neighbors = list(self.arc_graph.neighbors(robot_id))

            for neighbor_id in neighbors:
                if neighbor_id == robot_id:
                    continue

                if self.id_to_robots[neighbor_id].group_id != group.group_id:
                    continue

                sub_graph.add_node(neighbor_id)
                if not sub_graph.has_edge(robot_id, neighbor_id):
                    weight = self.arc_graph[robot_id][neighbor_id]['weight']
                    sub_graph.add_edge(robot_id, neighbor_id, weight=weight)

        # Calculate betweenness centrality
        betweenness_centrality = nx.betweenness_centrality(sub_graph, weight='weight')

        # This I measures betweenness centrality
        replace_leader = ad_leaders[0]
        max_iscore = -1.0

        for ad_leader in ad_leaders:
            bc_value = betweenness_centrality.get(ad_leader.robot_id, 0.0)
            iscore = (bc_value + 1) / (1 - (1 - ad_leader.fault_a) * (1 - ad_leader.fault_o))
            if iscore > max_iscore:
                replace_leader = ad_leader
                max_iscore = iscore

        group.leader = replace_leader
        ad_leaders.remove(replace_leader)
        group.ad_leaders = ad_leaders
